README
This is the README file for the pxruler package.

Pxruler provides a basic 'pixel ruler' for measuring pixel distances on
computer monitors.


NEWS
This is the NEWS file for pxruler.

------------------------------------------------------------------------------
v0.5.0 - 11 August 2022

Initial release.

------------------------------------------------------------------------------


build_debs.sh
#!/bin/sh

############################################################################
#  WARNING! WARNING! WARNING!                                              #
#                                                                          #
#  The contents of this file are automatically generated by the build      #
#  system. Do not make any changes directly to this file; such changes     #
#  will be silently ignored and overwritten!                               #
#                                                                          #
#  The template whence this file's contents were derived are in a file     #
#  of the same name with ".in" appended.                                   #
############################################################################

# build_debs.sh
#
#  Build Debian packages.
#
#   (C) Copyright 2021 Fred Gleason <fredg@paravelsystems.com>
#
#   This program is free software; you can redistribute it and/or modify
#   it under the terms of the GNU General Public License version 2 as
#   published by the Free Software Foundation.
#
#   This program is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#   GNU General Public License for more details.
#
#   You should have received a copy of the GNU General Public
#   License along with this program; if not, write to the Free Software
#   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
#

tar -zvxf lwmon-0.0.0.tar.gz
mv lwmon-0.0.0.tar.gz lwmon_0.0.0.orig.tar.gz

cd lwmon-0.0.0
dpkg-buildpackage -us -uc
cd ..

rm -rf lwmon-0.0.0


src/pxruler/pxruler.h
// pxruler.h
//
// GUI Ruler
//
//   (C) Copyright 2022 Fred Gleason <fredg@paravelsystems.com>
//
//   This program is free software; you can redistribute it and/or modify
//   it under the terms of the GNU General Public License version 2 as
//   published by the Free Software Foundation.
//
//   This program is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.
//
//   You should have received a copy of the GNU General Public
//   License along with this program; if not, write to the Free Software
//   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
//

#ifndef PXRULER_H
#define PXRULER_H

#include <QMainWindow>

#define PXRULER_USAGE "[--direction=vert|horz]"

class MainWidget : public QMainWindow
{
 Q_OBJECT;
 public:
  MainWidget();
  QSize sizeHint() const;

 protected:
  void paintEvent(QPaintEvent *e);

 private:
  Qt::Orientation d_orientation;
};


#endif  // PXRULER_H


src/pxruler/cmdswitch.h
// cmdswitch.h
//
// Process Command-Line Switches
//
//   (C) Copyright 2002-2022 Fred Gleason <fredg@paravelsystems.com>
//
//   This program is free software; you can redistribute it and/or modify
//   it under the terms of the GNU General Public License version 2 as
//   published by the Free Software Foundation.
//
//   This program is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.
//
//   You should have received a copy of the GNU General Public
//   License along with this program; if not, write to the Free Software
//   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
//

#ifndef CMDSWITCH_H
#define CMDSWITCH_H

#include <QList>
#include <QString>
#include <QStringList>

class CmdSwitch
{
 public:
  CmdSwitch(const QString &modname,const QString &usage);
  CmdSwitch(int argc,char *argv[],const QString &modname,
	    const QString &usage);
  unsigned keys() const;
  QString key(unsigned n) const;
  QString value(unsigned n) const;
  bool processed(unsigned n) const;
  void setProcessed(unsigned n,bool state);
  bool allProcessed() const;
  bool debugActive() const;

 private:
  QStringList switch_keys;
  QStringList switch_values;
  QList<bool> switch_processed;
  bool switch_debug;
};


#endif  // CMDSWITCH_H


src/pxruler/cmdswitch.cpp
// cmdswitch.cpp
//
// Process Command-Line Switches
//
//   (C) Copyright 2002-2022 Fred Gleason <fredg@paravelsystems.com>
//
//   This program is free software; you can redistribute it and/or modify
//   it under the terms of the GNU General Public License version 2 as
//   published by the Free Software Foundation.
//
//   This program is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.
//
//   You should have received a copy of the GNU General Public
//   License along with this program; if not, write to the Free Software
//   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
//

#include <stdio.h>
#include <syslog.h>
#include <stdlib.h>

#include <QCoreApplication>
#include <QMessageBox>

#include "cmdswitch.h"

CmdSwitch::CmdSwitch(const QString &modname,const QString &usage)
{
  switch_debug=false;

  QStringList args=qApp->arguments();

  for(int i=1;i<args.size();i++) {
    QString value=args.at(i);
    if(value=="--version") {
      printf("%s v%s \n",modname.toUtf8().constData(),VERSION);
      exit(0);
    }
    if(value=="--help") {
      printf("\n%s %s\n",modname.toUtf8().constData(),
	     usage.toUtf8().constData());
      exit(0);
    }
    if(value=="-d") {
      switch_debug=true;
    }
    QStringList f0=value.split("=",QString::KeepEmptyParts);
    if(f0.size()>=2) {
      if(f0.at(0).left(1)=="-") {
	switch_keys.push_back(f0.at(0));
	for(int i=2;i<f0.size();i++) {
	  f0[1]+="="+f0.at(i);
	}
	if(f0.at(1).isEmpty()) {
	  switch_values.push_back("");
	}
	else {
	  switch_values.push_back(f0.at(1));
	}
      }
      else {
	switch_keys.push_back(f0.join("="));
	switch_values.push_back("");
      }
      switch_processed.push_back(false);
    }
    else {
      switch_keys.push_back(value);
      switch_values.push_back("");
      switch_processed.push_back(false);
    }
  }
}


CmdSwitch::CmdSwitch(int argc,char *argv[],const QString &modname,
			 const QString &usage)
{
  switch_debug=false;

  for(int i=1;i<argc;i++) {
    QString value=QString::fromUtf8(argv[i]);
    if(value=="--version") {
      printf("%s v%s\n",modname.toUtf8().constData(),VERSION);
      exit(0);
    }
    if(value=="--help") {
      printf("\n%s %s\n",modname.toUtf8().constData(),
	     usage.toUtf8().constData());
      exit(0);
    }
    if(value=="-d") {
      switch_debug=true;
    }
    QStringList f0=value.split("=",QString::KeepEmptyParts);
    if(f0.size()>=2) {
      if(f0.at(0).left(1)=="-") {
	switch_keys.push_back(f0.at(0));
	for(int i=2;i<f0.size();i++) {
	  f0[1]+="="+f0.at(i);
	}
	if(f0.at(1).isEmpty()) {
	  switch_values.push_back("");
	}
	else {
	  switch_values.push_back(f0.at(1));
	}
      }
      else {
	switch_keys.push_back(f0.join("="));
	switch_values.push_back("");
      }
      switch_processed.push_back(false);
    }
    else {
      switch_keys.push_back(value);
      switch_values.push_back("");
      switch_processed.push_back(false);
    }
  }
}


unsigned CmdSwitch::keys() const
{
  return switch_keys.size();
}


QString CmdSwitch::key(unsigned n) const
{
  return switch_keys[n];
}


QString CmdSwitch::value(unsigned n) const
{
  return switch_values[n];
}


bool CmdSwitch::processed(unsigned n) const
{
  return switch_processed[n];
}


void CmdSwitch::setProcessed(unsigned n,bool state)
{
  switch_processed[n]=state;
}


bool CmdSwitch::allProcessed() const
{
  for(int i=0;i<switch_processed.size();i++) {
    if(!switch_processed[i]) {
      return false;
    }
  }
  return true;
}


bool CmdSwitch::debugActive() const
{
  return switch_debug;
}


src/pxruler/moc_pxruler.cpp
/****************************************************************************
** Meta object code from reading C++ file 'pxruler.h'
**
** Created by: The Qt Meta Object Compiler version 67 (Qt 5.9.7)
**
** WARNING! All changes made in this file will be lost!
*****************************************************************************/

#include "pxruler.h"
#include <QtCore/qbytearray.h>
#include <QtCore/qmetatype.h>
#if !defined(Q_MOC_OUTPUT_REVISION)
#error "The header file 'pxruler.h' doesn't include <QObject>."
#elif Q_MOC_OUTPUT_REVISION != 67
#error "This file was generated using the moc from 5.9.7. It"
#error "cannot be used with the include files from this version of Qt."
#error "(The moc has changed too much.)"
#endif

QT_BEGIN_MOC_NAMESPACE
QT_WARNING_PUSH
QT_WARNING_DISABLE_DEPRECATED
struct qt_meta_stringdata_MainWidget_t {
    QByteArrayData data[1];
    char stringdata0[11];
};
#define QT_MOC_LITERAL(idx, ofs, len) \
    Q_STATIC_BYTE_ARRAY_DATA_HEADER_INITIALIZER_WITH_OFFSET(len, \
    qptrdiff(offsetof(qt_meta_stringdata_MainWidget_t, stringdata0) + ofs \
        - idx * sizeof(QByteArrayData)) \
    )
static const qt_meta_stringdata_MainWidget_t qt_meta_stringdata_MainWidget = {
    {
QT_MOC_LITERAL(0, 0, 10) // "MainWidget"

    },
    "MainWidget"
};
#undef QT_MOC_LITERAL

static const uint qt_meta_data_MainWidget[] = {

 // content:
       7,       // revision
       0,       // classname
       0,    0, // classinfo
       0,    0, // methods
       0,    0, // properties
       0,    0, // enums/sets
       0,    0, // constructors
       0,       // flags
       0,       // signalCount

       0        // eod
};

void MainWidget::qt_static_metacall(QObject *_o, QMetaObject::Call _c, int _id, void **_a)
{
    Q_UNUSED(_o);
    Q_UNUSED(_id);
    Q_UNUSED(_c);
    Q_UNUSED(_a);
}

const QMetaObject MainWidget::staticMetaObject = {
    { &QMainWindow::staticMetaObject, qt_meta_stringdata_MainWidget.data,
      qt_meta_data_MainWidget,  qt_static_metacall, nullptr, nullptr}
};


const QMetaObject *MainWidget::metaObject() const
{
    return QObject::d_ptr->metaObject ? QObject::d_ptr->dynamicMetaObject() : &staticMetaObject;
}

void *MainWidget::qt_metacast(const char *_clname)
{
    if (!_clname) return nullptr;
    if (!strcmp(_clname, qt_meta_stringdata_MainWidget.stringdata0))
        return static_cast<void*>(this);
    return QMainWindow::qt_metacast(_clname);
}

int MainWidget::qt_metacall(QMetaObject::Call _c, int _id, void **_a)
{
    _id = QMainWindow::qt_metacall(_c, _id, _a);
    return _id;
}
QT_WARNING_POP
QT_END_MOC_NAMESPACE


src/pxruler/pxruler.cpp
// pxruler.cpp
//
// GUI Ruler
//
//   (C) Copyright 2022 Fred Gleason <fredg@paravelsystems.com>
//
//   This program is free software; you can redistribute it and/or modify
//   it under the terms of the GNU General Public License version 2 as
//   published by the Free Software Foundation.
//
//   This program is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.
//
//   You should have received a copy of the GNU General Public
//   License along with this program; if not, write to the Free Software
//   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
//

#include <QApplication>
#include <QPainter>

#include "cmdswitch.h"
#include "pxruler.h"

MainWidget::MainWidget()
  : QMainWindow(NULL,Qt::CustomizeWindowHint|Qt::WindowTitleHint|Qt::WindowCloseButtonHint|Qt::WindowStaysOnTopHint)
{
  d_orientation=Qt::Horizontal;

  CmdSwitch *cmd=new CmdSwitch("pxruler",PXRULER_USAGE);
  for(unsigned i=0;i<cmd->keys();i++) {
    if(cmd->key(i)=="--direction") {
      if(cmd->value(i).toLower()=="horz") {
	d_orientation=Qt::Horizontal;
      }
      else {
	if(cmd->value(i).toLower()=="vert") {
	  d_orientation=Qt::Vertical;
	}
	else {
	  fprintf(stderr,"pxruler: invalid direction\n");
	  exit(1);
	}
      }
      cmd->setProcessed(i,true);
    }
    if(!cmd->processed(i)) {
      fprintf(stderr,"pxruler: unrecognized option \"%s\"\n",
	      cmd->key(i).toUtf8().constData());
      exit(1);
    }
  }
  setWindowTitle("pxRuler");

  if(d_orientation==Qt::Horizontal) {
    setMinimumWidth(200);
    setMinimumHeight(sizeHint().height());
    setMaximumHeight(sizeHint().height());
  }
  else {
    setMinimumHeight(200);
    setMinimumWidth(sizeHint().width());
    setMaximumWidth(sizeHint().width());
  }
}


QSize MainWidget::sizeHint() const
{
  if(d_orientation==Qt::Horizontal) {
    return QSize(400,135);
  }
  return QSize(160,400);
}


void MainWidget::paintEvent(QPaintEvent *e)
{
  int h=size().height();
  int w=size().width();
  QPainter *p=new QPainter(this);

  p->setPen(palette().color(QPalette::WindowText));
  p->setFont(QFont(font().family(),24,QFont::Normal));


  if(d_orientation==Qt::Vertical) {
    p->translate(160,0);
    p->rotate(90);
    h=size().width();
    w=size().height();
  }

  p->drawText((w-p->fontMetrics().width(tr("Pixel Ruler")))/2,50,
	      tr("Pixel Ruler"));
  p->setFont(font());
  for(int i=0;i<w;i+=10) {
    p->drawLine(i,h-10,i,h);
  }
  for(int i=0;i<w;i+=50) {
    p->drawLine(i,h-20,i,h);
  }
  for(int i=0;i<w;i+=100) {
    p->drawLine(i,h-30,i,h);
    p->drawText(i+5,h-25,QString::asprintf("%dpx",i));
  }

  delete p;
}


int main(int argc,char *argv[])
{
  QApplication a(argc,argv);
  MainWidget *w=new MainWidget();
  w->show();

  return a.exec();
}
